### ISSUE TITLE
SPIN Rule Runs Multiple/Variable Number of Times for One Instantiation of Associated Class

### ISSUE THREAD

**Issue by gwcox66** [eclipse/rdf4j#231]
I'm using TopBraid Free Edition to create OWL ontologies with SPIN rules. I'm loading the ontology and SPIN rules into Sesame OpenRDF Workbench:

```
Application Name    OpenRDF Workbench
Version 4.1.2
Runtime Information
Operating System    Windows 8.1 6.3 (amd64)
Java Runtime    Oracle Corporation Java HotSpot(TM) 64-Bit Server VM (1.8.0_91)
Process User    Greg
Memory
Used    646 MB
Maximum 3463 MB
```
I've created a simple test ontology (RDF/SPIN file). It has one class with one SPIN rule and one datatype property.

```
CONSTRUCT {
    ?this BugReproduction:hasTimeStamp ?timeStamp .
}
WHERE {
    BIND (now() AS ?timeStamp) .
}
```
The SPIN rule adds an xsd:dateTime timestamp to the test instance of my test class. Here's the RDF for the entire test ontology including my class, datatype property, and rule (short):

````
    <?xml version="1.0"?>
<rdf:RDF
    xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
    xmlns:owl="http://www.w3.org/2002/07/owl#"
    xmlns:sp="http://spinrdf.org/sp#"
    xmlns:BugReproduction="http://www.disa.mil/dso/a2i/ontologies/PBSM/Interface/BugReproduction#"
    xmlns:rdfs="http://www.w3.org/2000/01/rdf-schema#"
    xmlns:spin="http://spinrdf.org/spin#"
    xmlns:spl="http://spinrdf.org/spl#"
    xmlns:arg="http://spinrdf.org/arg#"
    xmlns:xsd="http://www.w3.org/2001/XMLSchema#"
  xml:base="http://www.disa.mil/dso/a2i/ontologies/PBSM/Interface/BugReproduction">
  <owl:Ontology rdf:about="">
    <owl:imports rdf:resource="http://spinrdf.org/spl"/>
    <owl:versionInfo rdf:datatype="http://www.w3.org/2001/XMLSchema#string"
    >Created with TopBraid Composer</owl:versionInfo>
  </owl:Ontology>
  <owl:Class rdf:ID="TimeStampBug">
    <rdfs:subClassOf rdf:resource="http://www.w3.org/2002/07/owl#Thing"/>
    <spin:rule>
      <sp:Construct>
        <sp:templates rdf:parseType="Collection">
          <rdf:Description>
            <sp:object rdf:parseType="Resource">
              <sp:varName rdf:datatype="http://www.w3.org/2001/XMLSchema#string"
              >timeStamp</sp:varName>
            </sp:object>
            <sp:predicate>
              <owl:DatatypeProperty rdf:ID="hasTimeStamp"/>
            </sp:predicate>
            <sp:subject rdf:resource="http://spinrdf.org/spin#_this"/>
          </rdf:Description>
        </sp:templates>
        <sp:where rdf:parseType="Collection">
          <sp:Bind>
            <sp:expression>
              <sp:now/>
            </sp:expression>
            <sp:variable rdf:parseType="Resource">
              <sp:varName rdf:datatype="http://www.w3.org/2001/XMLSchema#string"
              >timeStamp</sp:varName>
            </sp:variable>
          </sp:Bind>
        </sp:where>
      </sp:Construct>
    </spin:rule>
  </owl:Class>
  <owl:DatatypeProperty rdf:about="#hasTimeStamp">
    <rdfs:range rdf:resource="http://www.w3.org/2001/XMLSchema#dateTime"/>
  </owl:DatatypeProperty>
</rdf:RDF>

<!-- Created with TopBraid -->`
```

So, I clear my SPIN repository in Sesame and use the workbench's Modify/Add command (with the "use base URI as context identifier" box unchecked which is important to avoid an issue already reported). Then I use a SPARQL Update query to create a test instance of my class:

```
`PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
PREFIX sxxicc: <http://www.disa.mil/dso/a2i/ontologies/PBSM/Interface/SXXIComplianceCheck#>
PREFIX owl: <http://www.w3.org/2002/07/owl#>
PREFIX sp: <http://spinrdf.org/sp#>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
PREFIX smf: <http://topbraid.org/sparqlmotionfunctions#>
PREFIX fn: <http://www.w3.org/2005/xpath-functions#>
PREFIX spl: <http://spinrdf.org/spl#>
PREFIX spin: <http://spinrdf.org/spin#>
PREFIX arg: <http://spinrdf.org/arg#>
PREFIX SXXIComplianceCheckIndividuals: <http://www.disa.mil/dso/a2i/ontologies/PBSM/Interface/SXXIComplianceCheckIndividuals#>
PREFIX sxxicci: <http://www.disa.mil/dso/a2i/ontologies/PBSM/Interface/SXXIComplianceCheckIndividuals#>
PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>
PREFIX bugs: <http://www.disa.mil/dso/a2i/ontologies/PBSM/Interface/BugReproduction#>
PREFIX bugsi: <http://www.disa.mil/dso/a2i/ontologies/PBSM/Interface/BugReproductionInstantiations#>

INSERT DATA {
    bugsi:aTimeStampBug_test1 a bugs:TimeStampBug .
}`
```

I then explore the resulting bugsi:aTimeStampBug_test1 individual, to find the SPIN:rule has run several times and produced several timestamps. The number of timestamps varies from test to test. Here's one example result:

![image](https://cloud.githubusercontent.com/assets/20132242/16771780/6791e282-4817-11e6-9b54-2436d2ed9f76.png)

The bottom line is that the SPIN rule runs multiple times for the one instance, and the number of times varies when I re-run this experiment.  I believe and have had another stackoverlow user confirm that the SPIN rules for a class should only run once per instance of that class (at instantiation time).  If this is correct, then this seems to be a bug which I'm duly reporting.

Thanks.

**Comment by jeenbroekstra** [comment]
At first glance this indeed looks like a bug (and apart from anything else, there may be quite a performance penalty involved here as well). Copying in @pulquero on this issue since he's the SPIN reasoner expert.

**Comment by gwcox66** [comment]
One further clue.  I re-ran this same test only changing the SPIN:rule to a SPIN:constructor.  The SPIN:constructor ran only once, producing but one timestamp.

![image](https://cloud.githubusercontent.com/assets/20132242/16785403/8620ee96-4853-11e6-8fef-415c088adc05.png)

**Comment by pulquero** [comment]
From my reading of the spec, it only says constructors must run once for an instance of a class. I believe rules and constraints should be expected to run more than once. Whenever you run the constraints, you would probably want to run the rules too. And you probably want to do this on any change to the statements. SpinSail proactively does this whereas TBC/Jena afaict require active intervention. This may lead to unanticipated behaviour when going from one to the other. Why did you choose a rule over a constructor? Or, what are you trying to achieve?

**Comment by gwcox66** [comment]
Yes, @pulquero I can use constructors instead.  I remain puzzled that a single instantiation (one triple in the SPARQL UPDATE query above) could results in a wildly-variable number of rule executions.   I re-ran the experiment this morning and got hundreds of time stamps (screen capture below).  
![image](https://cloud.githubusercontent.com/assets/20132242/16847104/ee605d36-49b4-11e6-8efd-d51653942140.png)

**Comment by gwcox66** [comment]
`SPIN:rule` is supposed to run every time the state of the instance changes.  I accidentally set up an infinite loop of changes, probably a classic mistake, but not a bug.

**Comment by pulquero** [comment]
I think you hit the nail on the head. That didnt occur to me but yes adding the timestamp means you are going to trigger the rules again. You should be able to use spin:rulePropertyMaxIterationCount to prevent this, in fact to quote the spec:

>  The property spin:rulePropertyMaxIterationCount can be attached to any sub-property of spin:rule to instruct the engine to execute certain rules a certain number of times. In particular this makes it possible to make sure that rules that create blank nodes are only executed once, avoiding problems with infinite loops.
